substrate/frame/contracts/src/wasm/mod.rs

```diff
@@ -410,25 +408,34 @@ impl<T: Config> Executable<T> for WasmBlob<T> {
 			.add_fuel(fuel_limit)
 			.expect("We've set up engine to fuel consuming mode; qed");
 
-		let exported_func = instance
-			.get_export(&store, function.identifier())
-			.and_then(|export| export.into_func())
-			.ok_or_else(|| {
-				log::error!(target: LOG_TARGET, "failed to find entry point");
-				Error::<T>::CodeRejected
-			})?;
-
-		if let &ExportedFunction::Constructor = function {
-			WasmBlob::<T>::increment_refcount(self.code_hash)?;
-		}
-
-		let result = exported_func.call(&mut store, &[], &mut []);
-		let engine_consumed_total = store.fuel_consumed().expect("Fuel metering is enabled; qed");
 		// Sync this frame's gas meter with the engine's one.
-		let gas_meter = store.data_mut().ext().gas_meter_mut();
-		gas_meter.charge_fuel(engine_consumed_total)?;
+		let process_result = |mut store: Store<Runtime<E>>, result| {
+			let engine_consumed_total =
+				store.fuel_consumed().expect("Fuel metering is enabled; qed");
+			let gas_meter = store.data_mut().ext().gas_meter_mut();
+			gas_meter.charge_fuel(engine_consumed_total)?;
+			store.into_data().to_execution_result(result)
+		};
 
-		store.into_data().to_execution_result(result)
+		match instance.start(&mut store) {
+			Ok(instance) => {
+				let exported_func = instance
+					.get_export(&store, function.identifier())
+					.and_then(|export| export.into_func())
+					.ok_or_else(|| {
+						log::error!(target: LOG_TARGET, "failed to find entry point");
+						Error::<T>::CodeRejected
+					})?;
+
+				if let &ExportedFunction::Constructor = function {
```

Comment:
What if the `start` function internally calls the `deploy` export, for `call` (and vice versa)? I suppose nothing bad should happen as the contract code is completely untrusted right?

Reply:
This is possible and there is nothing wrong with it. The start function is metered and doesn't have any special semantics.

Reply:
Thinking about it again: If the contract traps or terminates etc. in the start function we should do no further call into the blob. Needs some thinking rework.